### ISSUE TITLE
Xamarin.Android task cannot find NDK with latest Azure image(s)

### ISSUE THREAD

**Issue by Maxgamerboy1** [xamarin/xamarin-android#7233]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

Azure pipeline

### Description

Making use of the Xamarin.Android task we have been able to build and deploy Xamarin.Android applications. New images were released which seems to have affected this task; the NDK path is no longer found when executing the "_ResolveMonoAndroidSdks" step resulting in a build failure.

Reviewing the latest releases for these images, I see there was work around NDK paths, symlinks, and environment variable changes. 
https://github.com/actions/virtual-environments/issues/5879
https://github.com/actions/virtual-environments/issues/5930

The image we're using is `macOS-latest`, here are the old and new logs showing the "_ResolveMonoAndroidSdks" step resolving the NDK path:
Old working image (20220724.1)
![Screenshot 2022-08-04 at 12 37 45](https://user-images.githubusercontent.com/25750702/182837912-a6b4aaaf-08a9-48a0-9027-aabb500f7d2d.png)

New broken image (20220801.1)
![Screenshot 2022-08-04 at 12 38 40](https://user-images.githubusercontent.com/25750702/182838071-be4977dd-12b8-4f63-bef4-812e20e945fe.png)

Do we know if the Xamarin.Android task needs to be updated to reflect this?

### Steps to Reproduce

1. Make use of the Azure Pipeline Xamarin.Android task for a Xamarin.Android application using the `macOS-latest` image.
2. See the task eventually fails with `error XA5104: Could not locate the Android NDK. Please make sure the Android NDK is installed in the Android SDK Manager, or if using a custom NDK path, please ensure the $(AndroidNdkDirectory) MSBuild property is set to the custom path.`

**Comment by jonathanpeppers** [comment]
Started a new one, let's see what it does: https://dev.azure.com/jonathanpeppers/boots/_build/results?buildId=33&view=results

**Comment by jonathanpeppers** [comment]
I think it also worked: https://dev.azure.com/jonathanpeppers/boots/_build/results?buildId=33&view=logs&j=a82d9e72-1de4-5bd6-c188-90296135479e&t=65080fd3-055e-56ae-0221-9a17a9cb5386&l=365

**Comment by Maxgamerboy1** [comment]
Thanks for the response! Our builds started failing this morning, ones running yesterday were still using the old image, due to the staggered rollout.

I do see your build is using MSBuild directly, whereas I'm using this badger for building my Xamarin Android project
https://docs.microsoft.com/en-gb/azure/devops/pipelines/tasks/build/xamarin-android?view=azure-devops

**Comment by jonathanpeppers** [comment]
It seems like the bots should still have r23 and r24, can you add some `ls` commands to a build to find out if they are there?

**Comment by jonathanpeppers** [comment]
Mine is also using `XamarinAndroidVersion: 13.0.0-0`, which it downloaded and installed.

**Comment by GreatBarrier86** [comment]
Instead of the suggested mitigation of installing r21, is there a way to change the project to just use the latest version? I'm obviously having the same problem too.

**Comment by jonathanpeppers** [comment]
I found something that prevents the selection of r25: https://github.com/xamarin/xamarin-android-tools/pull/190

Versions with Android NDK r24 or lower should work, though. As long as one of those is installed.

**Comment by GreatBarrier86** [comment]
According to [https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md](url), the following NDKs are installed.
21.4.7075529
23.2.8568313
24.0.8215888

Wouldn't one of these be auto-selected? Or is there a configuration problem on my manfiest/csproj side? It just seems like a manual step shouldn't be necessary

**Comment by DevenCC** [comment]
Hey there, following the mitigation proposed in [this thread](https://github.com/actions/virtual-environments/issues/5930), I've added this task prior to my Android build task and at least the NDK is now present.
```
    - task: CmdLine@2
      displayName: "XA5104 Workaround"
      inputs:
        script: '${ANDROID_HOME}/tools/bin/sdkmanager --install "ndk;21.4.7075529"'
```

This might not be an elegant permanent solution, since as @GreatBarrier86 stated, I also don't understand why 21.4.7075529 is not found automatically. However, this could help unblock those failing pipelines in the meantime at least.

**Comment by GreatBarrier86** [comment]
@DevenCC, did you have any success referencing a later version of the NDK that, according to that link, is also installed on the image? I tried doing the r24 and had no success.

**Comment by DevenCC** [comment]
@GreatBarrier86  I tried the same task with 24.0.8215888 and it worked fine as well. 
I am assuming here that using any of the ndk versions stating in the [MacOS-latest virtual environments](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md) would work.
I do not know if any other specific version would work however.

**Comment by jtorvald** [comment]
I'm experiencing the same issue. This week we had one re-run on devops and that one magically succeeded while the previous one failed. All the builds after kept failing. The temp-fix from @DevenCC helps to get a successful build.

**Comment by jonathanpeppers** [comment]
Is someone able to share a `.binlog` of a failing build? And if you have a working example to compare against, that would also be helpful. Thanks!

You can build with `-bl:nameyourlog.binlog`, and then archive the file during your build.

**Comment by jtorvald** [comment]
@jonathanpeppers I tried to make one but I can't find where the binlog is placed. I copied the whole source and output directory but can't find it.

**Comment by mikequ-taggysoft** [comment]
Our App Center Xamarin.Forms Android builds just broke with the same error, too. We don't use any MSBuild task script or specify any specific NDK path/version in our Xamarin project.  Things always built fine with just the source code in our DevOps repo.

I wonder why it's not using whatever is the latest on the build agent after this update. Is there a setting in the Xamarin.Forms project that we have to configure?

**Comment by jtorvald** [comment]
@mikequ-taggysoft look at the temporary fix from @DevenCC with `task: CmdLine@2`. Copy that block into your pipeline.yml before the task `task: XamarinAndroid@1`
That will make sure the NDK gets installed and your build will succeed.

**Comment by mikequ-taggysoft** [comment]
@jtorvald We're not using Azure Pipelines for the project. We build it on-commit or manually via App Center which is connected to the repo.

**Comment by jtorvald** [comment]
@mikequ-taggysoft sorry, I skipped over that. I guess you can do the same in AppCenter with a post-clone hook script and to add this line to it:

```
$ANDROID_HOME/tools/bin/sdkmanager --install "ndk;21.4.7075529"
```

Documentation about post-clone hook can be found [here](https://docs.microsoft.com/en-us/appcenter/build/custom/scripts/#post-clone)

**Comment by mikequ-taggysoft** [comment]
@jtorvald Thanks! We do use post clone scripts and adding that command to it did resolve the issue.

Is there any way to find out exactly which part of our Xamarin project depends on NDK v21 (or the Xamarin.Android framework itself does?) Is it possible to use the latest version somehow?

**Comment by GreatBarrier86** [comment]
Something else is strange here. I dont have any NDK version installed on my local build machine and i have no trouble building my app at all. If it works fine for me without the NDK, why is it necessary on the azure build machine? Is it something MacOS specific?

Also, i agree with @mikequ-taggysoft. If we can eliminate the specific version requirement, that would be ideal.

![image](https://user-images.githubusercontent.com/29482393/183413250-2ffd8e78-bd4a-4122-baf6-fee466ac40a7.png)

**Comment by mikequ-taggysoft** [comment]
In my case, I believe my project requires NDK for sure because the release build profile uses AOT. But I can't see anywhere in the project config where it specifies a specific version of NDK.

**Comment by GreatBarrier86** [comment]
@dellis1972 , i bet that's it. i dont do release builds locally.

**Comment by jonathanpeppers** [comment]
I tried again this morning, and it seems to be working for me:

https://dev.azure.com/jonathanpeppers/boots/_build/results?buildId=34&view=logs&j=a82d9e72-1de4-5bd6-c188-90296135479e&t=65080fd3-055e-56ae-0221-9a17a9cb5386

```
XamarinAndroidVersion: 13.0.0-0
...
Android NDK: /Users/runner/Library/Android/sdk/ndk/24.0.8215888/
```

This is CI for [boots](https://github.com/jonathanpeppers/boots), and it does basically:
```
dotnet tool install --global boots --version 1.1.0.712-preview2
boots --preview Xamarin.Android
```

Can anyone comment on the version of Xamarin.Android where you're having this problem? Or share a log? Does updating Xamarin.Android solve the issue?

**Comment by jonathanpeppers** [comment]
> @jonathanpeppers I tried to make one but I can't find where the binlog is placed. I copied the whole source and output directory but can't find it.

@jtorvald sometimes on macOS, I found MSBuild doesn't put the `.binlog` where you expect unless you do something like:
```
-bl:$(System.DefaultWorkingDirectory)/msbuild.binlog
```

**Comment by jtorvald** [comment]
@jonathanpeppers thank. I will give it a spin to see if I can send you a binlog

**Comment by Maxgamerboy1** [comment]
> It seems like the bots should still have r23 and r24, can you add some `ls` commands to a build to find out if they are there?

Sorry this took a while. I ran the following command: `echo ANDROID_NDK_HOME; ls $(ANDROID_NDK_HOME); echo "sdk path"; ls /Users/runner/Library/Android/sdk`
Which resulted in this output:

Generating script.
Script contents:
echo ANDROID_NDK_HOME; ls /Users/runner/Library/Android/sdk/ndk/25.0.8775105; echo "sdk path"; ls /Users/runner/Library/Android/sdk
========================== Starting Command Output ===========================
/bin/bash /Users/runner/work/_temp/d01ea05a-ce5e-4d55-88c4-e784972de9da.sh
**ANDROID_NDK_HOME**
CHANGELOG.md
NOTICE
NOTICE.toolchain
README.md
build
meta
ndk-build
ndk-gdb
ndk-lldb
ndk-stack
ndk-which
package.xml
prebuilt
python-packages
shader-tools
simpleperf
source.properties
sources
toolchains
wrap.sh
**sdk path**
build-tools
cmake
cmdline-tools
emulator
extras
licenses
ndk
patcher
platform-tools
platform-tools-2
platforms
tools

As we can see, it seems the NDK is available on the machine. After a bit more investigation, I was able to supply additional build arguments to the task:  `msbuildArguments: '/p:AndroidNdkDirectory="$(ANDROID_NDK_HOME)"'

The task then succeeds to find NDK 25.0.8775105 as specified by the build image
https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md#environment-variables-2

I was unable to get a binlog, as supplying this argument resolved our issue. It is still strange how specifying the NDK wasn't required before.

**Comment by jtorvald** [comment]
@jonathanpeppers I've a binlog. Can I share it somewhere semi-private? Could I drop you an email at your Gmail or create a private ticket as MS somehow that you can pick up?

**Comment by jonathanpeppers** [comment]
@jtorvald you can email at the address on my GitHub profile, thanks.

**Comment by jonathanpeppers** [comment]
Ok, this log message is a bit different than what I'm seeing!
```
Skipping NDK in '/Users/runner/Library/Android/sdk/ndk/24.0.8215888': version 24.0.8215888 is out of the accepted range (major version must be between 16 and 21
Skipping NDK in '/Users/runner/Library/Android/sdk/ndk/23.2.8568313': version 23.2.8568313 is out of the accepted range (major version must be between 16 and 21
Skipping NDK in '/Users/runner/Library/Android/sdk/ndk/25.0.8775105': version 25.0.8775105 is out of the accepted range (major version must be between 16 and 21
```

**Comment by mikequ-taggysoft** [comment]
@jonathanpeppers I use `Xamarin.Forms` and all packages are the latest. It still gives me that error. I sent App Center build logs to your email. This is the Xamarin build configuration in my csproj:

```
<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
<DebugSymbols>false</DebugSymbols>
<DebugType>portable</DebugType>
<Optimize>true</Optimize>
<OutputPath>bin\Release</OutputPath>
<ErrorReport>prompt</ErrorReport>
<WarningLevel>4</WarningLevel>
<AndroidManagedSymbols>true</AndroidManagedSymbols>
<AndroidUseSharedRuntime>false</AndroidUseSharedRuntime>
<AotAssemblies>true</AotAssemblies>
<!-- Hybrid mode is required to strip IL codes from DLLs in APKs, leaving only class/method names for reflection -->
<AndroidAotMode>Hybrid</AndroidAotMode>
<!-- These two aot arguments reduce APK size -->
<AndroidAotAdditionalArguments>no-write-symbols,nodebug</AndroidAotAdditionalArguments>
<EnableLLVM>false</EnableLLVM>
<AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
<BundleAssemblies>false</BundleAssemblies>
<EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
<AndroidUseAapt2>true</AndroidUseAapt2>
<AndroidCreatePackagePerAbi>false</AndroidCreatePackagePerAbi>
<AndroidSupportedAbis>arm64-v8a</AndroidSupportedAbis>
</PropertyGroup>

```

**Comment by Maxgamerboy1** [comment]
> I was unable to get a binlog, as supplying this argument resolved our issue. It is still strange how specifying the NDK wasn't required before.

I thought this resolved our issue as it got past the resolve step, yet a different error occurred within the `_BuildApkEmbed` task; "error XA5105: Toolchain utility 'armv7a-linux-androideabi16-clang' for target Arm was not found. Tried in path: '/Users/runner/Library/Android/sdk/ndk/25.0.8775105/toolchains/llvm/prebuilt/darwin-x86'"

Downloading and examining the r25 NDK on my local machine I am seeing that `armv7a-linux-androideabi16-clang` does not exist. The closest thing that does exist lives under "[...]/toolchains/llvm/prebuilt/darwin-x86_64/bin/armv7a-linux-androideabi19-clang". Notice we're under the /bin folder and that the closest -clang is 19. 16 doesn't exist.

My guess as to where this new issue lies it two-fold:
1. https://github.com/xamarin/xamarin-android/blob/51acc438ef547c3ad3b2a8113a014c3539c231cd/src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/WithClang.cs#L111 where we should append the `bin` folder (the darwin folder is based on the machine's architecture so I don't think this should change).
2. Updating the minimum level [ https://github.com/xamarin/xamarin-android/blob/51acc438ef547c3ad3b2a8113a014c3539c231cd/src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/WithClang.cs#L41 , https://github.com/xamarin/xamarin-android/blob/0ab3db079d43d47ad1bc522e28e65204542fea60/build-tools/scripts/XABuildConfig.cs.in#L21 ]

Yet, I'm unsure how to make these changes only apply to the r25 NDK. Should I also perhaps look to open a separate issue, as this is more to do with using the r25 NDK rather than trying to find it.

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, you're using a preview version of Xamarin.Android, and according to your pipeline, you have a step for installing it. How about trying without installing your version of it and just use whatever the latest default on the build image?

**Comment by jonathanpeppers** [comment]
@GreatBarrier86 when looking at @jtorvald's log:

https://github.com/xamarin/xamarin-android/issues/7233#issuecomment-1208332820

It has the same Xamarin.Android version as me, but it's acting as if it has a really old `Xamarin.Android.Tools.Android.Sdk.dll` in comparison to the `Xamarin.Android.Common.targets` file. It makes me think there is something very odd about the AzDO machine's image.

Mine must be working, because it installed the `.pkg` after and all the files are updated appropriately? I'm unsure, I'm trying to find a project where I can reproduce this.

Anyone hitting this, have you tried using `boots` to install Xamarin.Android? Does that fix it?

**Comment by jtorvald** [comment]
@jonathanpeppers 

Adding this before the XamarinAndroid@1 step also solves the issue. You have a binlog and azure build log in your email.
```
- task: CmdLine@2
  displayName: "XA5104 Workaround"
  inputs:
    script: 'dotnet tool install --global boots --version 1.1.0.712-preview2 && boots --preview Xamarin.Android'
```

**Comment by jonathanpeppers** [comment]
I think I reproduced the problem here: https://github.com/jonathanpeppers/maui-profiling/pull/20

Going to save Xamarin.Android files as artifacts, to figure out what is going on...

**Comment by jonathanpeppers** [comment]
There is definitely something up with the AzDO image. I downloaded the suspect `.dll` and it is ancient:

![image](https://user-images.githubusercontent.com/840039/183522433-8c62a2c7-6266-4162-963c-44b14e7482c0.png)

After I install my own `.pkg` I get:

![image](https://user-images.githubusercontent.com/840039/183522393-0fc72713-9703-40f2-acf6-aa0c5f1df5ae.png)

**Comment by jonathanpeppers** [comment]
I filed an issue here: https://github.com/actions/runner-images/issues/6038

Looking to find the right person to message at Microsoft internally.

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, was the closure of #7245 supported to resolve this issue? Just wanted to follow up because it's still an issue. Obviously the workaround continues to be fine so I have no problem using it for now.

Current agent version: '2.206.1'
Current image version: '20220815.1'

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, was the closure of #6038 supposed to resolve this issue? Wanted to follow up since it still appears to be an issue, even though they said the rollout of the update would take 3-4 days from early last week.

Current image version: '20220815.1'
Current agent version: '2.206.1'

**Comment by jonathanpeppers** [comment]
@GreatBarrier86 if you have a newer `.binlog` from your build, can you check the value of `$(XamarinAndroidVersion)` now?

**Comment by GreatBarrier86** [comment]
With the workarounds in place or not?

**Comment by jonathanpeppers** [comment]
We're basically just checking if they updated Xamarin.Android on the image.

**Comment by GreatBarrier86** [comment]
Sorry but I’m not familiar with how to produce and publish the binlog file. Do you have info on how?

**Comment by jonathanpeppers** [comment]
Maybe just pass `-v:diag` to your build (to get diagnostic logs), and then search for `XamarinAndroidVersion`.

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, while i didnt find that variable in the build log, i did see the following related to the NDK. I thought this was what was to be resolved in that other thread.

I did attach the build log for that step though, since i could have missed where the Android Version was mentioned

![image](https://user-images.githubusercontent.com/29482393/186030182-c36e88a1-2bb3-4542-81e8-9744696cae98.png)
[16.txt](https://github.com/xamarin/xamarin-android/files/9397999/16.txt)

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, i did notice i can build the app without the workaround NDK line now, so that might have fixed your issue.

**Comment by Maxgamerboy1** [comment]
Ah! Thanks for the nudge! I'll give it a go.

**Comment by rubeonline** [comment]
Same issue with vmImage: windows-2019:

<img width="652" alt="image" src="https://user-images.githubusercontent.com/5664460/186186002-06c99ed8-db0c-495c-9f6e-f047332aa9ef.png">

**Comment by jonathanpeppers** [comment]
@rubeonline you will need to use VS 2022 to get the latest Xamarin.Android. I assume that image has VS 2019?

**Comment by rubeonline** [comment]
On 4. July my build ran also on VS2019 without any issues. I assume that nothing changes if i use the same image?

**Comment by jonathanpeppers** [comment]
@rubeonline if they dropped an Android NDK from the image, and then there isn't an NDK that this old version of Xamarin.Android supports -- you'll probably have an issue. In that case you may need to install an older NDK yourself, or newer Xamarin.Android.

**Comment by Maxgamerboy1** [comment]
> @Maxgamerboy1, i did notice i can build the app without the workaround NDK line now, so that might have fixed your issue.

@GreatBarrier86 I gave this a go with the latest mac os image 20220822.4, and it's still failing with the same error
`/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(1954,3): error XA5104: Could not locate the Android NDK. Please make sure the Android NDK is installed in the Android SDK Manager, or if using a custom NDK path, please ensure the $(AndroidNdkDirectory) MSBuild property is set to the custom path.`

I noticed the `$(ANDROID_NDK_HOME)` is populated correctly. In the XamarinAndroid task we be passing `/p:AndroidNdkDirectory="$(ANDROID_NDK_HOME)"` variable to the `msbuildArguments` property? We haven't done this before this error occurred, we just let that task find the NDK automatically.

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, how is the resolution coming for lack of support for Android 13? Even the new '20220822.4' image version is having this issue.

**Comment by jonathanpeppers** [comment]
Android 13 support is only in Xamarin.Android 13.x or higher, which comes in Visual Studio 2022. I'm not sure what purpose someone would need the VS 2019 image.

You can install Xamarin.Android yourself if you need a newer version.

**Comment by GreatBarrier86** [comment]
I'm not sure what you're referring to. I'm using macos-latest, which according to here (https://github.com/actions/runner-images/blob/main/images/macos/macos-11-Readme.md), has 2022 installed by default.

You mentioned in this discussion here (https://github.com/xamarin/xamarin-android/issues/7245), that it was because of the broken image. Are you saying this should be fixed by now?

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, have you had any success without that line yet?

![image](https://user-images.githubusercontent.com/29482393/187218676-bda363c3-3dd8-4e2b-a59d-f0d7626c5d72.png)

**Comment by Maxgamerboy1** [comment]
@GreatBarrier86 I'm still having the same issue; `error XA5104: Could not locate the Android NDK.`... I just want to make sure we're doing the same thing. I'm using the Xamarin.Android Task to build;
```
==============================================================================
Task         : Xamarin.Android
Description  : Build an Android app with Xamarin
Version      : 1.208.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/build/xamarin-android
==============================================================================
```
Using this is what's failing for me.

I did find that this task allows me to pass in the NDK directory, as it exposes `msbuildArguments` as a parameter, and passing this to it;
```
/p:AndroidNdkDirectory="$(ANDROID_NDK_HOME)"
```
Populates the NDK correctly
![Screenshot 2022-08-30 at 10 58 26](https://user-images.githubusercontent.com/25750702/187408391-4c25135b-cc6f-4b6b-acd2-b56a0d3436a1.png)

But this causes a different issue; `error XA5105: Toolchain utility 'armv7a-linux-androideabi16-clang' for target Arm was not found. Tried in path: "/Users/runner/Library/Android/sdk/ndk/25.0.8775105/toolchains/llvm/prebuilt/darwin-x86"`
Looping back to my previous comment https://github.com/xamarin/xamarin-android/issues/7233#issuecomment-1208444711

**Comment by Maxgamerboy1** [comment]
This is what we define;
```
  - task: XamarinAndroid@1
    displayName: 'Build ${{ parameters.ProjectPathMain }}'
    inputs:
      projectFile: ${{ parameters.ProjectPathMain }}
      outputDirectory: '$(build.binariesdirectory)/${{ parameters.BuildConfiguration }}'
      configuration: ${{ parameters.BuildConfiguration }}
      msbuildVersionOption: latest
      msbuildArguments: '/p:AndroidNdkDirectory="$(ANDROID_NDK_HOME)"'
      jdkOption: 'JDKVersion'
      createAppPackage: true
```
This is what was used which resulted in that error; `error XA5105: Toolchain utility 'armv7a-linux-androideabi16-clang' for target Arm was not found. Tried in path: "/Users/runner/Library/Android/sdk/ndk/25.0.8775105/toolchains/llvm/prebuilt/darwin-x86"`

**Comment by jonathanpeppers** [comment]
@Maxgamerboy1 can you use this to get a `.binlog` of your build:

```yaml
msbuildArguments: -bl:$(Build.ArtifactStagingDirectory)msbuild.binlog
```

Then save this file as an artifact, and share it?

You might consider using the plain `MSBuild` yaml task instead of `XamarinAndroid`. I'm not sure who created it, or if it is maintained. It seems like the `MSBuild` task would do everything and you just need to use the `-t:SignAndroidPackage` target?

**Comment by Maxgamerboy1** [comment]
> you can email at the address on my GitHub profile, thanks.

@jonathanpeppers i've emailed you a binlog from an errored build
i'll look to use MSBuild directly tomorrow

**Comment by GreatBarrier86** [comment]
@jonathanpeppers, i shared the XamarinAndroidVersion above in this post here: https://github.com/xamarin/xamarin-android/issues/7233#issuecomment-1223252344. Is that satisfactory to get what you need for this issue? Not the NDK one, but the issue of 2022 not supporting Android 13.

**Comment by jonathanpeppers** [comment]
@Maxgamerboy1 @GreatBarrier86 you both have Xamarin.Android 12.x, and you need to use 13.x. What Azure DevOps pool are you using?

They supposedly rolled out an update that contains Xamarin.Android 13.x with: https://github.com/actions/runner-images/pull/6045

**Comment by GreatBarrier86** [comment]
Do you want a new log with `-v:diag` in it?

**Comment by jonathanpeppers** [comment]
If you see `XamarinAndroidVersion` says 12.x with `macos-latest`, you might consider filing an issue here: https://github.com/actions/runner-images/issues

Or we could reopen: https://github.com/actions/runner-images/issues/6038

**Comment by GreatBarrier86** [comment]
> You might consider using the plain `MSBuild` yaml task instead of `XamarinAndroid`. I'm not sure who created it, or if it is maintained. It seems like the `MSBuild` task would do everything and you just need to use the `-t:SignAndroidPackage` target?

@jonathanpeppers, do you recommend using the regular MSBuild task instead of Xamarin.Android? I created a new pipeline to develop Xamarin.Android and it put in both the Xamarin.Android step AND the MSBuild one. If we should be using the other one, that's fine, but it would be good to know.
![image](https://user-images.githubusercontent.com/29482393/187794767-2cf11afa-3544-4676-838a-e7d9aeeb8e7e.png)

**Comment by Maxgamerboy1** [comment]
> @Maxgamerboy1 @GreatBarrier86 you both have Xamarin.Android 12.x, and you need to use 13.x. What Azure DevOps pool are you using?
> 
> They supposedly rolled out an update that contains Xamarin.Android 13.x with: [actions/runner-images#6045](https://github.com/actions/runner-images/pull/6045)

I'm using the Azure Pipelines pool, running the macOS-latest image, specifically that errored build used image version 20220829.1. 

I changed to use the MSBuild task for my android project and it still fails to find the NDK, and it continues to resolve Xamarin.Android 12.0.0.3... 

Testing with a brand new project, it works without issue, though i'm not seeing where we would specify i want to use Xamarin.Android 13.x/12.x

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, according to here (https://github.com/actions/runner-images/issues/6038#issuecomment-1233429689), `macos-latest` doesn't actually have the latest soft. Can you get it to work with `macos-12`?

**Comment by Maxgamerboy1** [comment]
@GreatBarrier86 thanks for pointing that out, it works using macos-12!
I'll be switching over to MSBuild task instead of the XamarinAndroid one too.

I'd be happy to mark this as resolved

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, any reason you're using the MSBuild task over the XamarinAndroid one? just curious myself. I'd like to use whichever one is recommended.

**Comment by jonathanpeppers** [comment]
> @GreatBarrier86 no real reason, i'm seeing Pepper's builds making use of it, and with the revelation that we don't know who is maintaining that task, I think it's a smart move to hop over the MSBuild task

I don't think I use the `XamarinAndroid` .yaml task, where did you see that? If someone is working on lately, it isn't the Xamarin.Android team, so I'm not sure about its status.

Personally, I don't see what value it has over the regular MSBuild task. If there is some value, we should just put that logic in Xamarin.Android/.NET 6 instead?

**Comment by GreatBarrier86** [comment]
Unless i misunderstand, he's saying he does NOT use Xamarin.Android and uses MSBuild instead.

**Comment by Maxgamerboy1** [comment]
> Unless i misunderstand, he's saying he saw that you do NOT use Xamarin.Android, but instead use MSBuild.

Yes, that's correct. Sorry for the confusion; XamarinAndroid.yml - I've stopped using, MSBuild - I've now switched over to use :D

**Comment by salgiza** [comment]
Could someone post an example of how the MSBuild task should be configured to get the same result as this basic XamarinAndroid task?

![CleanShot 2022-09-02 at 14 36 44](https://user-images.githubusercontent.com/6630306/188145171-63b82e18-2bf8-41e4-80bb-b9cc60af66c2.png)

We are having the same problem and I don't really mind using one or the other as long as it works, but all the docs for Azure DevOps use the XamarinAndroid task, and I'm totally unfamiliar with the MSBuild task 😢

**Comment by Maxgamerboy1** [comment]
@salgiza sure! Although, see my response here https://github.com/xamarin/xamarin-android/issues/7233#issuecomment-1234315475 the issue is relating more to the build image itself rather than the task used to build.

to switch over, in your case, it should look something like this;
```yaml
  - task: MSBuild@1
    inputs:
      solution: '**/*Droid*.csproj'
      configuration: '$(buildConfiguration)'
      msbuildArguments: '-p:OutDir="$(outputDirectory)" -t:SignAndroidPackage'
```

**Comment by salgiza** [comment]
Thanks a lot @Maxgamerboy1 , with all the messages in the thread, I had missed the one that said that changing the vmImage from macos-latest to macos-12 should fix the problem. Seeing the comments from @jonathanpeppers, though, I decided to do both (change the vmImage, and migrate from the XamarinAndroid task to MSBuild), and our pipeline is working again!

**Comment by jonathanpeppers** [comment]
Ok, so I think we've solved this one -- the `macos-12` image has Xamarin.Android 13.x, which supports up to Android NDK 24.

Let me know if I missed anything outstanding here, thanks!

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, one thing I noticed, and I'm not sure what I should be using here, but the Xamarin.Android step's msbuild line shows `/t:PackageForAndroid` instead of `/t:SignAndroidPackage`. My next step after this is the `Android signing` task, so I'm not sure what I should be using here.

![image](https://user-images.githubusercontent.com/29482393/188961541-c9499644-7886-48c8-ab0a-736933271da9.png)

**Comment by Maxgamerboy1** [comment]
Ah, sorry, didn't know you were doing signing in a separate step. Yes, you'll need to use the `PackageForAndroid` target in that case.

**Comment by GreatBarrier86** [comment]
I certainly don't have to sign in a different step. Does the msbuild task also zipalign? If so, i don't see why i can't use the msbuild task for both.

**Comment by Maxgamerboy1** [comment]
If using the `/t:SignAndroidPackage` it should zipalign, then keysign for you, here's a snippet from one of my old builds showing the _Sign step;
```
2022-09-05T14:26:12.3762630Z _Sign:
2022-09-05T14:26:12.3763980Z   /Users/runner/Library/Android/sdk/build-tools/32.0.0/zipalign -p 4 "obj/release/android/bin/com.[---].apk" "/Users/runner/work/1/b/release//com.[---]-Signed.apk" 
2022-09-05T14:26:13.1999960Z   /Library/Java/JavaVirtualMachines/Temurin-Hotspot-11.jdk/Contents/Home/bin/java -jar /Library/Frameworks/Xamarin.Android.framework/Libraries/xbuild/Xamarin/Android/apksigner.jar sign --ks "/Users/runner/.local/share/Xamarin/Mono for Android/debug.keystore" --ks-pass pass:android --ks-key-alias androiddebugkey --key-pass pass:android --min-sdk-version 26 --max-sdk-version 30  /Users/runner/work/1/b/release/com.[---]-Signed.apk 
```
you'll also note prior to this step, MSBuild creates a debug keystore under the `_CreateAndroidDebugSigningKey` step.

essentially, if you wanna zipalign and sign using your own keys, use `/t:PackageForAndroid`, else let the `/t:SignAndroidPackage` do it for you

**Comment by GreatBarrier86** [comment]
@Maxgamerboy1, I'm leaving this here for anyone else that happens to run across this same question. You can pass in MSBuild arguments to sign the APK file, regardless of whether its your key or not. Maybe it will help someone in the future because i did need to research how to do this.

`/p:OutputPath=$(build.binariesdirectory)/$(BuildConfiguration)/$(TestOutputFolder) /t:SignAndroidPackage /p:AndroidKeyStore=true /p:AndroidSigningKeyAlias=$(KeystoreAlias) /p:AndroidSigningKeyPass=$(CertificatePassword) /p:AndroidSigningKeyStore=$(Build.SourcesDirectory)/MyKeystore.file /p:AndroidSigningStorePass=$(CertificatePassword)`
